sprinkle/low-level/context.py

```diff
@@ -0,0 +1,127 @@
+library experience; the context object is sort of like an instance of a
+library that has some information bound within it that is particular to the
+current running task.  E.g. det.get_trial_id() would be a strange library call
+but context.get_trial_id() is very natural.
+
+Goals for context objects in the Sprinkle API:
+
+ - There should only be one context class for all things PyTorchTrial, and
+   only one for all things Keras, etc.  The context class for PyTorchTrial and
+   the context class for Keras don't have to be the same though.
+
+ - The same context object must be usable in both local training and cluster
+   training environments.  Obviously if it exposes cluster-training-specific
+   extensions during cluster training, those extensions do not need to work in
+   local training environments, but any user code which does not depend on
+   such environment-specific extensions should run equally well in either
+   environment.
+
+ - The same context object must be usable in both training and non-training
+   workloads.  The sprinkle API for training should feel analagous to the
+   sprnkle API for distributed batch inference and to distributed test set
+   metrics, even though the calls the user makes will differ.
+
+ - Backwards compatibility: the existing Trial APIs need to still work, so
+   those TrialContext classes should remain unchanged.
+"""
+
+# GENERIC CONTEXT
+# Useful for custom workloads and is the basis for the framework contexts.
+context = det.GenericContext()
+
+# distributed data: applies to any workload; training or inference, local or cluster.
+# (same as the DistributedContext right now)
+context.distributed.get_rank()
+context.distributed.get_size()
+context.distributed.get_local_rank()
+context.distributed.get_num_agents()
+
+# rendezvous data: info about machines and IPs that are participating in distributed jobs.
+# This is configured automatically by the rendezvous info layer in Determined, but can also be
+# configured automatically through an environment variable if you want to use Determined's dtrain
+# features outside of Determined.
+context.rendezvous.get_num_nodes()
+context.rendezvous.get_node_rank()
+context.rendezvous.get_slots_per_node()
+
+# master data: applies to any job running on-cluster, of any job that you have
+# chosen to connect to the master with context.connect_to_master()
+context.connect_to_master(...)  # args similar to Determined() object
+context.master.get_addr()
+context.master.get_port()
+context.master.get_tls()
+
+# cluster data: applies to any job running on-cluster
+context.cluster.get_container_id()
+context.cluster.get_container_gpus()
+context.cluster.get_slot_ids()
+
+# training-related data; only applies to `det e create` workloads
+context.training.get_experiment_config()
+context.training.get_data_config()
+context.training.get_experiment_id()
+context.training.get_trial_id()
+context.training.get_global_batch_size()
```

Comment:
`context.training.get_global_batch_size()` is there for backwards compatibility with the TrialContext object, and it would raise an error if you decided *not* to set `global_batch_size` in the hyperparameters but then you called that function anyway.

Reply:
@rb-determined-ai Can you put this explanation in the spec?